src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataDictionaryFormatter.cs

```diff
@@ -0,0 +1,363 @@
+                        break;
+
+                    case ObjectType.String:
+                        response = messagePackReader.ReadString();
+                        break;
+
+                    case ObjectType.Char:
+                        response = messagePackReader.ReadChar();
+                        break;
+
+                    case ObjectType.Guid:
+                        response = options.Resolver.GetFormatterWithVerify<Guid>().Deserialize(ref messagePackReader, options);
+                        break;
+
+                    case ObjectType.CreationPolicy:
+                        response = (CreationPolicy)messagePackReader.ReadByte();
+                        break;
+
+                    case ObjectType.Type:
+                        response = options.Resolver.GetFormatterWithVerify<TypeRef?>().Deserialize(ref messagePackReader, options).Resolve();
+                        break;
+
+                    case ObjectType.TypeRef:
+                        response = options.Resolver.GetFormatterWithVerify<TypeRef?>().Deserialize(ref messagePackReader, options);
+                        break;
+
+                    case ObjectType.Enum32Substitution:
+                        TypeRef? enumType = options.Resolver.GetFormatterWithVerify<TypeRef?>().Deserialize(ref messagePackReader, options);
+                        int rawValue = options.Resolver.GetFormatterWithVerify<int>().Deserialize(ref messagePackReader, options);
+                        response = new LazyMetadataWrapper.Enum32Substitution(enumType!, rawValue);
+                        break;
+
+                    case ObjectType.TypeSubstitution:
+                        TypeRef? typeRef = options.Resolver.GetFormatterWithVerify<TypeRef?>().Deserialize(ref messagePackReader, options);
+                        response = new LazyMetadataWrapper.TypeSubstitution(typeRef!);
+                        break;
+
+                    case ObjectType.TypeArraySubstitution:
+                        IReadOnlyList<TypeRef?> typeRefArray = options.Resolver.GetFormatterWithVerify<IReadOnlyList<TypeRef?>>().Deserialize(ref messagePackReader, options);
+                        response = new LazyMetadataWrapper.TypeArraySubstitution(typeRefArray!, options.CompositionResolver());
+                        break;
+
+                    case ObjectType.TypeLess:
+                        var typeLessData = options.Resolver.GetFormatterWithVerify<byte[]>().Deserialize(ref messagePackReader, options);
+                        response = MessagePackSerializer.Typeless.Deserialize(typeLessData);
+                        break;
+
+                    default:
+                        throw new NotSupportedException(string.Format(CultureInfo.CurrentCulture, Strings.UnsupportedFormat, objectType));
+                }
+
+                return response;
+            }
+        }
+
+        private enum ObjectType : byte
+        {
+            Null,
```

Comment:
I know the original serialization code had this, but as msgpack has a native representation for nil, I suggest we just use that.